zendesk/importer.go

```diff
@@ -0,0 +1,187 @@
+	CustomFields        []interface{} `json:"custom_fields"`
+	SatisfactionRating  interface{}   `json:"satisfaction_rating"`
+	SharingAgreementIds []interface{} `json:"sharing_agreement_ids"`
+	Fields              []interface{} `json:"fields"`
+	FollowupIds         []interface{} `json:"followup_ids"`
+	TicketFormID        int64         `json:"ticket_form_id"`
+	AllowChannelback    bool          `json:"allow_channelback"`
+	AllowAttachments    bool          `json:"allow_attachments"`
+	GeneratedTimestamp  int           `json:"generated_timestamp"`
+}
+
+type Via struct {
+	Channel string `json:"channel"`
+	Source  Source `json:"source"`
+}
+
+type Source struct {
+	From interface{} `json:"from"`
+	To   interface{} `json:"to"`
+	Rel  interface{} `json:"rel"`
+}
+
+type BulkImporter struct {
+	userName   string       // userName to login as admin to zendesk
+	apiToken   string       // token to authenticate the user
+	client     *http.Client // http client to connect zendesk
+	maxRetries uint64       // max API retries in case of 429, before returning error
+	baseURL    string       // zendesk api url
+	retryCount uint64       // number of retry count made for current data
+}
+
+// NewBulkImporter initialize bulk importer to write bulk tickets to zendesk
+func NewBulkImporter(userName, apiToken, domain string, maxRetries uint64) *BulkImporter {
+	return &BulkImporter{
+		//url:        fmt.Sprintf("https://%s.zendesk.com/api/v2/imports/tickets/create_many", domain),
+		client:     newHTTPClient(),
+		userName:   userName,
+		apiToken:   apiToken,
+		baseURL:    fmt.Sprintf("https://%s.zendesk.com", domain),
+		maxRetries: maxRetries,
+	}
+}
+
+// Write buffer data to zendesk
+func (b *BulkImporter) Write(ctx context.Context, records []sdk.Record) error {
+	bufferedTicket, err := parseRecords(records)
+	if err != nil {
+		return fmt.Errorf("unable to parse the records %w", err)
+	}
+
+	req, err := http.NewRequestWithContext(
+		ctx,
+		http.MethodPost,
+		fmt.Sprintf("%s/api/v2/imports/tickets/create_many", b.baseURL),
+		bytes.NewBuffer(bufferedTicket),
+	)
+
+	if err != nil {
+		return fmt.Errorf("unable to send to zendesk server %w", err)
+	}
+
+	req.Header.Set("Content-Type", "application/json; charset=UTF-8")
+	req.Header.Add("Authorization", "Basic "+basicAuth(b.userName, b.apiToken))
+
+	resp, err := b.client.Do(req)
+	if err != nil {
+		return fmt.Errorf("unable to fetch response from zendesk server %w", err)
+	}
+
+	defer resp.Body.Close()
+	// Validation for httpStatusCode 429 - Too many Requests, Retry value after `93s`
+	if resp.StatusCode == http.StatusTooManyRequests {
+		// NOTE: https://developer.zendesk.com/documentation/ticketing/using-the-zendesk-api/best-practices-for-avoiding-rate-limiting/#catching-errors-caused-by-rate-limiting
+		retryValue, err := strconv.ParseInt(resp.Header.Get("Retry-After"), 10, 64)
+		if err != nil {
+			return fmt.Errorf("unable to get retry value: %w", err)
```

Comment:
If the value is invalid, maybe it's better to just use a default value. The fact that we cannot parse the header is probably not important enough that we do not retry at all.